src/expr/node_converter.cpp

```diff
@@ -0,0 +1,254 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Node converter utility
+ */
+
+#include "expr/node_converter.h"
+
+#include "expr/attribute.h"
+
+using namespace cvc5::kind;
+
+namespace cvc5 {
+
+NodeConverter::NodeConverter(bool forceIdem) : d_forceIdem(forceIdem) {}
+
+Node NodeConverter::convert(Node n)
+{
+  if (n.isNull())
+  {
+    return n;
+  }
+  Trace("nconv-debug") << "NodeConverter::convert: " << n << std::endl;
+  NodeManager* nm = NodeManager::currentNM();
+  std::unordered_map<Node, Node>::iterator it;
+  std::vector<TNode> visit;
+  TNode cur;
+  visit.push_back(n);
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+    it = d_cache.find(cur);
+    Trace("nconv-debug2") << "convert " << cur << std::endl;
+    if (it == d_cache.end())
+    {
+      d_cache[cur] = Node::null();
+      Assert(d_preCache.find(cur) == d_preCache.end());
+      Node curp = preConvert(cur);
+      // if curp = cur, set null to avoid infinite loop
```

Comment:
I don't understand this comment - couldn't there be a legitimate conversion that would not change some nodes? I understand why you don't want to push `curp` to `visit` in such a case, but I don't see why the entry in the cache should point to null.

Reply:
I'll comment, preCache should be null when you don't convert, otherwise it should point to the node that you should follow to get the pre-rewritten form.